### DISCUSSION TITLE
How to make an redirect?

### DISCUSSION THREAD

**Discussion by the-man-with-a-golden-mind** [freshframework/fresh#511] [Q&A] [answered]
Hello, I have to make redirect for auth process. 
Is it possible with fresh? I can not find anything in the Fresh documentation about that.

Best,
MM

**Comment by lj-n** [comment]
You can return a Response with a Location header in your handler functions:
```ts
export function handler(req: Request): Response {
  // ...
  const headers = new Headers({
    location: new URL(req.url).origin,
  });
  return new Response(null, {
    status: 302,
    headers,
  });
}
```

Hope this helps.

**Comment by lucacasonato** [comment, marked as answer]
If you have a fully qualified URL, you can use `Response.redirect`.

```ts
export function handler(req: Request): Response {
  return Response.redirect("https://example.com", 307);
}
```

If you want to redirect to a relative path on the current domain:

```ts
export function handler(req: Request): Response {
  return new Response("", {
    status: 307,
    headers: { Location: "/my/new/relative/path" },
  });
}
```

**Reply by ruzmetovJob** [reply]
hello, can you tell me how to get the user back.
for example, I have a page in which the script has worked and now the user must return to where he came from

**Reply by TranquilMarmot** [reply]
This is great! It's a bummer that I had to dig around to find an answer here 😕

It seems like this docs page could use some more examples (I would think a redirect is a pretty common thing, yeah?)

https://fresh.deno.dev/docs/concepts/middleware

**Reply by marvinhagemeister** [reply]
Agree, made an issue to track that https://github.com/denoland/fresh/issues/1607

**Comment by ruzmetovJob** [comment]
@lucacasonato hello, can you tell me how to get the user back.
for example, I have a page in which the script has worked and now the user must return to where he came from

**Reply by vixalien** [reply]
what about something like: `new Response.redirect(req.url)`

**Reply by leesei** [reply]
SSR server is so like Web 1.0 😆.
You can try seting the return URL in cookies or in the query param of the page where the script run.

**Comment by chopfitzroy** [comment]
I am having some issues with this, I have a route: `routes/singout.tsx` which has the following content:

```tsx
import { HandlerContext, Handlers } from "$fresh/server.ts";
import { MiddlewareState } from "./_middleware.ts";

export const handler: Handlers = {
  GET(_, ctx: HandlerContext<unknown, MiddlewareState>) {
    const pb = ctx.state.pb;
    pb.authStore.clear();
    return new Response(undefined, {
      status: 307,
      headers: { Location: "/" },
    });
  },
};
```

But when I navigate to `localhost:8000/signout` it does not redirect me back to `localhost:8000/` the network tab looks like this:

![image](https://user-images.githubusercontent.com/2978850/209921732-4341c924-e1b7-4a4c-a8d4-46a29148c3eb.png)

I have also tried the technique recommended in this [thread](https://github.com/denoland/fresh/issues/428):

```tsx
 const url = new URL(req.url);
 url.pathname ="/";
 return Response.redirect(url, 307);
```

 But got the same result. Code is all opensource and can be viewed [here](https://github.com/chopfitzroy/coffeeandcode.app).

**Reply by sylc** [reply]
@chopfitzroy  The issue you are having is not with the redirect but with your middleware that does not pass properly the status code.
The below I believe should work.
 
in https://github.com/chopfitzroy/coffeeandcode.app/blob/main/routes/_middleware.ts#L32 

```diff
-  return new Response(resp.body, { ...resp, headers });
+  return new Response(resp.body, { status: resp.status, headers });
```

**Comment by iJKTen** [comment]
I have this code and I see a blank page and it does not redirect. 
```
export const handler: Handlers<ActionData> = {
  GET(req, ctx) {
    const cookies = getCookies(req.headers);
    const loggedInEmail = cookies?.trivial || null;

    if (!loggedInEmail) {
      return new Response(null, {
        status: 401,
        headers: { Location: "/login" },
      });
    }

    return ctx.render();
  },
};

```

I do not have any middleware logic.

**Reply by marvinhagemeister** [reply]
To redirect you have to respond with a status code in the 300 - 399 range. See https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#redirection_messages

**Comment by iJKTen** [comment]
Why would this line error out 
```
return Response.redirect(`${origin}/login`, 401);
```
with the error `RangeError: Invalid redirect status code.`?

**Reply by PhoneDroid** [reply]
To redirect you have to respond with a status code 
in the `300 - 399` range - the **[Redirect Messages]** -  
for example with `307` for temporary redirects.

```ts
return Response.redirect(`${origin}/login`,307);
```

[Redirect Messages]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#redirection_messages
